src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -367,43 +370,37 @@ internal FunctionActivityTracker(IOptionsMonitor<AppServiceOptions> appServiceOp
                     _metricsPublisher = metricsPublisher;
                 }
 
-                StartActivityTimer();
+                _activityFlushCurrSecond = _functionActivityFlushInterval;
+                _activityTimer = new Timer(TimerFlush, null, _activityTimerIntervalMS, _activityTimerIntervalMS);
+
+                _logger = logger;
             }
 
             internal IEventGenerator MetricsEventGenerator { get; private set; }
 
-            private void StartActivityTimer()
+            internal void TimerFlush(object state)
             {
-                Task.Run(
-                    async () =>
+                try
+                {
+                    // we do this every second
+                    RaiseMetricsPerFunctionEvent();
+
+                    // only raise function actity events when the interval expires
+                    if (_activityFlushCurrSecond >= _functionActivityFlushInterval)
                     {
-                        try
-                        {
-                            int currentSecond = _functionActivityFlushInterval;
-                            while (!_etwTaskCancellationSource.Token.IsCancellationRequested)
-                            {
-                                RaiseMetricsPerFunctionEvent();
-
-                                if (currentSecond >= _functionActivityFlushInterval)
-                                {
-                                    RaiseFunctionMetricEvents();
-                                    currentSecond = 0;
-                                }
-                                else
-                                {
-                                    currentSecond = currentSecond + 1;
-                                }
-
-                                await Task.Delay(TimeSpan.FromSeconds(1), _etwTaskCancellationSource.Token);
-                            }
-                        }
-                        catch (TaskCanceledException)
```

Comment:
Unhandled exceptions that weren't of this type cause this task to fail silently, meaning the metrics pump would then no longer be running. That leads to a memory leak because _runningFunctions is no longer being cleared of completed invocations.